firedrake/variational_solver.py

```diff
@@ -15,12 +15,28 @@
            "NonlinearVariationalSolver"]
 
 
+def check_pde_args(F, J, Jp):
+    if not isinstance(F, (ufl.Form, slate.slate.TensorBase)):
+        raise TypeError("Provided residual is a '%s', not a Form or Slate Tensor" % type(F).__name__)
+    if len(F.arguments()) != 1:
+        raise ValueError("Provided residual is not a linear form")
+    if not isinstance(J, (ufl.Form, slate.slate.TensorBase)):
+        raise TypeError("Provided Jacobian is a '%s', not a Form or Slate Tensor" % type(J).__name__)
+    if len(J.arguments()) != 2:
+        raise ValueError("Provided Jacobian is not a bilinear form")
+    if Jp is not None and not isinstance(Jp, (ufl.Form, slate.slate.TensorBase)):
+        raise TypeError("Provided preconditioner is a '%s', not a Form or Slate Tensor" % type(Jp).__name__)
+    if Jp is not None and len(Jp.arguments()) != 2:
+        raise ValueError("Provided preconditioner is not a bilinear form")
+
+
 class NonlinearVariationalProblem(object):
     r"""Nonlinear variational problem F(u; v) = 0."""
 
     def __init__(self, F, u, bcs=None, J=None,
                  Jp=None,
-                 form_compiler_parameters=None):
+                 form_compiler_parameters=None,
+                 is_linear=False):
```

Comment:
Seems a bit strange to have an `is_linear` argument. Wouldn't it be possible to simply check for instances of `LinearVariationalProblem`? Otherwise, let's add `is_linear` to the docstring and specify why one might set this to be `True`

Reply:
Thanks! The form style checks (`A == b` or `F == 0`) were actually missing in `NonlinearVariationalProblem.`  Cleaned up.